### ISSUE TITLE
X11 apps do not prompt for access to the microphone

### ISSUE THREAD

**Issue by JeffFessler** [XQuartz/XQuartz#245]
I am running a command-line app (Julia language) that tries to record using the system built-in microphone.

It works fine in a Mac Terminal, because Terminal is smart enough to ask for permission to use the mic.
But running the same app under XQuartz does not work because XQuartz does not generate a prompt
to the user to ask for permission to use the mic, and, strangely, XQuartz does not show up under the list
of apps in the "Security and Privacy" section of System Preferences, so I cannot manually give XQuartz
permission to use the mic.

Now as I write this, I realize it is unclear which app should be requesting permission.
I am running an xterm under XQuartz, and in that xterm I am launching Julia and running
some Julia code that in turn calls some libportaudio library.  It works in Mac's Terminal
so it feels like it should work in xterm too.  Suggestions?

The specific Julia app is this one: https://github.com/JuliaAudio/PortAudio.jl/issues/96

**Comment by jeremyhu** [comment]
That link is to a bug report in another project.  I don't see a link to an application I can install to reproduce this issue.  Do you have an application that I can easily build or install that will show off this issue?  If so, I can dig into it.

**Comment by jeremyhu** [comment]
> from an Xterm

This needs to be expanded on.

1. If you launch the xterm from Terminal.app, does it work?
2. If you launch the xterm from XQuartz.app's Applications menu, does it work?
3. If you launch the xterm from xinit (eg: by adding 'xterm' to ~/.xinitrc.d/launch_an_xterm.sh and making it executable), does it work?

**Comment by JeffFessler** [comment]
> * If you launch the xterm from Terminal.app, does it work?

Interestingly, in this case it records fine.  I had given Terminal permission to use the mic long ago, and apparently those permissions are propagated to the xterm launched from within it (which I had never tried before) and then in turn to the Julia job run in the xterm.  Cool!

> * If you launch the xterm from XQuartz.app's Applications menu, does it work?

That is my usual use, and that is the case that does not work that and that prompted this issue.


>  If you launch the xterm from xinit

I did even not have a directory `~/.xinitrc.d/` so first I tried running xinit (from an xterm) without doing any of that and xinit automatically started a xterm and recording did not work.  Then I tried adding the (executable) file in that directory, as suggested, and again starting xinit (from an xterm) which gave me a new xterm, and recording did not work in it either.  The man page for `xinit` discusses `~/.xinitrc` file, but does not discuss that director `~/.xinitrc.d/` so I am not as confident about this test.

So in summary it seems like xterm does not ask for permissions to use the mic no matter how it is started from within the Xterm ecosystem.

You've already given me a helpful work around because at least I can now start the xterm from a Terminal and have it work.  (You might wonder why not just run from within Terminal?  Well, I've used Xwindows for more than 30 years and my muscle memory is trained with the "focus follows mouse" mode...  If Terminal had that mode I would likely switch to it!)

**Comment by jeremyhu** [comment]
It has to do with what cohort the process is in.  When it is spawned from Terminal.app (or something below Terminal.app), Terminal is the responsible process, so thank you for confirming that.

When launched from the XQuartz.app menu, XQuartz should be the responsible process.  I'd expect that to work.

I was expecting that last case to be the one that failed because I thought it would be associated with a coalition for the LaunchAgent.

I'll ig into this when I get a chance...

**Comment by jeremyhu** [comment]
Also, when you get a chance, please file a report with Apple's Feedback Assistant.  Reference this bug report there, and report back with the FB number here.

**Comment by jeremyhu** [comment]
And be sure to include a sysdiagnose taken immediately after the failure

**Comment by jeremyhu** [comment]
Please verify this as fixed in 2.8.2 beta4 or later:
    https://www.xquartz.org/releases/XQuartz-2.8.2_beta4.html

**Comment by JeffFessler** [comment]
Yes!  I tested it with Julia on XQ 2.8.2 beta4 (with Mojave 10.14.6) and it worked.
It properly asked the OS for permission to use the mic, and then recorded audio correctly.  And now X11 shows up in the privacy settings under Sys. Pref.  Thanks!

> And be sure to include a sysdiagnose taken immediately after the failure

Sorry, I don't know what that means, and it never actually "failed" in the sense of causing an app crash.  It just silently recorded a 0 waveform for the designated recording interval.

**Comment by jeremyhu** [comment]
The FB was more for me to get the data from you.  I don't really need it now that it's fixed ;)

Glad it's working for you.  Thanks for reporting back.
